src/public/js/rs.js

```diff
@@ -45,13 +45,214 @@ var RS = (function() {
+    var lhs;
+    var delta;
+    var shiftedTau;
+    var shiftedGamma;
+    var scaledTau;
+    var scaledGamma;
+    var DThreshold;
+    for (var i = 0; i <= n - k; i++) {
+      // Compute the left-hand side of the key equation
+      lhs = UTIL.polynomialMult(sigmas[i], syndromePolyPlusOne);
+
+      // Delta is the coefficient of z^(i+1)
+      delta = lhs[(lhs.length - 1) - (i + 1)];
+
+      // Use tau and gamma to update sigma and omega:
+      shiftedTau = UTIL.mergeTypedArrays(taus[i], new Uint8Array([0])); // z * tau
+      shiftedGamma = UTIL.mergeTypedArrays(gammas[i], new Uint8Array([0])); // z * gamma
+      scaledTau = UTIL.polynomialScale(shiftedTau, delta); // delta * z * tau
+      scaledGamma = UTIL.polynomialScale(shiftedGamma, delta); // delta * z * gamma
+      sigmas[i + 1] = UTIL.polynomialAdd(sigmas[i], scaledTau);
+      omegas[i + 1] = UTIL.polynomialAdd(omegas[i], scaledGamma);
+
+      // Follow the B-M iteration rules to update tau and gamma:
+      DThreshold = (i + 1) / 2;
+      if (delta == 0 || Ds[i] > DThreshold || (Ds[i] == DThreshold && !Bs[i])) {
+        // RULE A
+        taus[i + 1] = UTIL.mergeTypedArrays(taus[i], new Uint8Array([0]));
+        gammas[i + 1] = UTIL.mergeTypedArrays(gammas[i], new Uint8Array([0]));
+      } else {
+        // RULE B
+        var invDelta = UTIL.multInv(delta);
+        taus[i + 1] = UTIL.polynomialScale(sigmas[i], invDelta);
+        gammas[i + 1] = UTIL.polynomialScale(omegas[i], invDelta);
+
+        // Update the D tracker
+        Ds[i + 1] = i + 1 - Ds[i];
+
+        // If the equality case triggered, turn the B flag off
+        if (Ds[i] == DThreshold) {
+          Bs[i + 1] = false;
+        }
+      }
+    }
+
+    return {
+      errorLocationPoly: sigmas[n - k],
+      errorMagnitudePoly: omegas[n - k]
+    };
+  }
+
+  // Performs a Chien search on an error location polynomial to find the set of
+  // error locations.
+  // Input errorLocationPoly is a big-endian array representing a polynomial,
+  // i.e. [1,2,3] -> x^2 + 2x + 3
+  // Output is a set of integers in GF(256), each of which is an error location, GEN^power
+  // -> i.e. taking log3 of an element in the set would give you the error position,
+  // the power at which the error occurred.
+  function chienSearch(errorLocationPoly) {
+    var xis = [];
+    for (var v = 0; v < 256; v++) {
+      var res = UTIL.polynomialEval(errorLocationPoly, v);
+      if (res == 0) {
+        xis.push(UTIL.multInv(v));
+      }
+    }
+    return xis;
+  }
+
+  // Performs Forney's Formula on a set of error locations and an error magnitude
+  // polynomial to find the set of error magnitudes.
+  // Input errorLocations is a set of integers in GF(256)
+  // Input errorMagnitudePoly is a big-endian array representing a polynomial,
+  // i.e. [1,2,3] -> x^2 + 2x + 3
+  // Output errorMagnitudes is a set of integers in GF(256)
+  function forneysFormula(errorLocations, errorMagnitudePoly) {
```

Comment:
Can we call the second one the errorEvaluatorPoly so it isn't confused with a polynomial holding the magnitude of the errors. I think it isn't really equivalent to the error magnitudes just used to solve for them?